enclave/core/sgx/entropy.c

```diff
@@ -3,24 +3,14 @@
 
 #include <openenclave/bits/safecrt.h>
 #include <openenclave/enclave.h>
-#include <openenclave/internal/random.h>
-
-/*
- * MBEDTLS links this function definition when MBEDTLS_ENTROPY_HARDWARE_ALT
- * is defined in the MBEDTLS config.h file. This is the sole source of entropy
- * for MBEDTLS. All other MBEDTLS entropy sources are disabled since they don't
- * work within enclaves.
- */
-int mbedtls_hardware_poll(
-    void* data,
-    unsigned char* output,
-    size_t len,
-    size_t* olen)
-{
-    int ret = -1;
-    unsigned char* p = output;
+#include <openenclave/internal/entropy.h>
+#include <openenclave/internal/rdrand.h>
 
-    OE_UNUSED(data);
+/* TODO: This should use RDSEED instead. See issue #242. */
+oe_result_t oe_get_entropy(void* output, size_t len)
+{
+    oe_result_t ret = OE_UNEXPECTED;
+    unsigned char* p = (unsigned char*)output;
```

Comment:
Can you also take a pass at cleaning up this function?
Do we really need to use memcpy_s here or just do a uint64_t volatile write?
We don't do any memcpy_s for the left over bytes.

Reply:
i'm not sure what you mean by a uint64_t volatile write. If you mean something like:
```C
while (n--)
{
    uint64_t x = oe_rdrand();
    *((uint64_t*) p) = x;
    p += sizeof(uint64_t);
}
```
That's really dangerous since p isn't guaranteed to be aligned to 8 bytes, so it can crash. You need a memcpy-like function like `memcpy_(s)` or the `while(...) { *p++ = *q++ }` loop.

Reply:
Good point about alignment.
We need to change line 40 to use memcpy_s